File: scripts/analyze_trial_outputs.py
#!/usr/bin/env python3
"""
Analyze UFFS trial run outputs - compare C++ vs Rust outputs for exact parity.

Usage:
    python scripts/analyze_trial_outputs.py <trial_dir>
    python scripts/analyze_trial_outputs.py docs/trial_runs/d_disk

Expected files in trial_dir:
    - cpp_d.txt (or cpp_<drive>.txt) - C++ reference output
    - rust_new_d.txt (or rust_new_<drive>.txt) - Rust with new tree algo

The script will:
1. Parse both files and extract all paths
2. Compare line-by-line for exact match
3. Report any differences with context
4. Analyze patterns in differences (if any)
"""

import sys
import os
from pathlib import Path
from collections import defaultdict
import re


def parse_uffs_output(filepath: Path) -> list[dict]:
    """Parse UFFS tab-separated output into list of records."""
    records = []
    with open(filepath, 'r', encoding='utf-8', errors='replace') as f:
        for line_num, line in enumerate(f, 1):
            line = line.rstrip('\n\r')
            if not line:
                continue
            # Tab-separated: path, type, size, descendants, date, ...
            parts = line.split('\t')
            if len(parts) >= 4:
                records.append({
                    'line_num': line_num,
                    'path': parts[0],
                    'type': parts[1] if len(parts) > 1 else '',
                    'size': parts[2] if len(parts) > 2 else '',
                    'descendants': parts[3] if len(parts) > 3 else '',
                    'raw': line
                })
    return records


def normalize_path(path: str) -> str:
    """Normalize path for comparison."""
    return path.lower().replace('\\', '/').rstrip('/')


def compare_outputs(cpp_file: Path, rust_file: Path) -> dict:
    """Compare C++ and Rust outputs, return detailed comparison."""
    print(f"\n{'='*70}")
    print(f"Comparing:")
    print(f"  C++:  {cpp_file}")
    print(f"  Rust: {rust_file}")
    print(f"{'='*70}")

    cpp_records = parse_uffs_output(cpp_file)
    rust_records = parse_uffs_output(rust_file)

    print(f"\nRecord counts:")
    print(f"  C++:  {len(cpp_records):,}")
    print(f"  Rust: {len(rust_records):,}")
    print(f"  Diff: {len(rust_records) - len(cpp_records):+,}")

    # Build path sets for comparison
    cpp_paths = {normalize_path(r['path']): r for r in cpp_records}
    rust_paths = {normalize_path(r['path']): r for r in rust_records}

    cpp_only = set(cpp_paths.keys()) - set(rust_paths.keys())
    rust_only = set(rust_paths.keys()) - set(cpp_paths.keys())
    common = set(cpp_paths.keys()) & set(rust_paths.keys())

    print(f"\nPath comparison:")
    print(f"  Common paths:  {len(common):,}")
    print(f"  C++ only:      {len(cpp_only):,}")
    print(f"  Rust only:     {len(rust_only):,}")

    if len(cpp_paths) > 0:
        match_pct = len(common) / len(cpp_paths) * 100
        print(f"  Match rate:    {match_pct:.4f}%")

    # Check for attribute differences in common paths
    attr_diffs = []
    for path in common:
        cpp_rec = cpp_paths[path]
        rust_rec = rust_paths[path]
        if cpp_rec['size'] != rust_rec['size'] or cpp_rec['descendants'] != rust_rec['descendants']:
            attr_diffs.append({
                'path': path,
                'cpp_size': cpp_rec['size'],
                'rust_size': rust_rec['size'],
                'cpp_desc': cpp_rec['descendants'],
                'rust_desc': rust_rec['descendants']
            })

    if attr_diffs:
        print(f"\n⚠️  Attribute differences in common paths: {len(attr_diffs)}")
        print(f"\nFirst 10 attribute differences:")
        for diff in attr_diffs[:10]:
            print(f"  {diff['path']}")
            print(f"    Size: C++={diff['cpp_size']} vs Rust={diff['rust_size']}")
            print(f"    Desc: C++={diff['cpp_desc']} vs Rust={diff['rust_desc']}")

    # Report missing paths
    if cpp_only:
        print(f"\n❌ Paths in C++ but NOT in Rust (first 20):")
        for p in sorted(cpp_only)[:20]:
            print(f"  {p}")
        if len(cpp_only) > 20:
            print(f"  ... and {len(cpp_only) - 20} more")

    if rust_only:
        print(f"\n❌ Paths in Rust but NOT in C++ (first 20):")
        for p in sorted(rust_only)[:20]:
            print(f"  {p}")
        if len(rust_only) > 20:
            print(f"  ... and {len(rust_only) - 20} more")

    # Perfect match check
    if not cpp_only and not rust_only and not attr_diffs:
        print(f"\n✅ PERFECT MATCH - All {len(common):,} paths match exactly!")
        return {'status': 'perfect', 'common': len(common)}

    return {
        'status': 'differences',
        'common': len(common),
        'cpp_only': len(cpp_only),
        'rust_only': len(rust_only),
        'attr_diffs': len(attr_diffs)
    }


def main():
    if len(sys.argv) < 2:
        print(__doc__)
        sys.exit(1)

    trial_dir = Path(sys.argv[1])
    if not trial_dir.exists():
        print(f"Error: Directory not found: {trial_dir}")
        sys.exit(1)

    # Find output files
    cpp_files = list(trial_dir.glob("cpp_*.txt"))
    rust_new_files = list(trial_dir.glob("rust_new_*.txt"))

    print(f"Trial directory: {trial_dir}")
    print(f"Found {len(cpp_files)} C++ files, {len(rust_new_files)} Rust (new) files")

    # Match by drive letter
    for cpp_file in sorted(cpp_files):
        drive = cpp_file.stem.replace('cpp_', '')
        rust_file = trial_dir / f"rust_new_{drive}.txt"
        if rust_file.exists():
            compare_outputs(cpp_file, rust_file)
        else:
            print(f"\n⚠️  No matching Rust file for {cpp_file.name}")


if __name__ == "__main__":
    main()

File: scripts/analyze_trial_outputs.rs
#!/usr/bin/env rust-script
//! Analyze UFFS trial run outputs - compare C++ vs Rust outputs for exact parity.
//!
//! Usage:
//!   rust-script scripts/analyze_trial_outputs.rs <trial_dir>
//!   rust-script scripts/analyze_trial_outputs.rs docs/trial_runs/d_disk
//!
//! Expected files in trial_dir:
//!   - cpp_d.txt (or cpp_<drive>.txt) - C++ reference output
//!   - rust_new_d.txt (or rust_new_<drive>.txt) - Rust with new tree algo
//!
//! The script will:
//! 1. Parse both files and extract all paths
//! 2. Compare for exact match
//! 3. Report any differences with context
//! 4. Analyze patterns in differences (if any)
//!
//! ```cargo
//! [dependencies]
//! ```

use std::collections::{HashMap, HashSet};
use std::env;
use std::fs::{self, File};
use std::io::{BufRead, BufReader};
use std::path::{Path, PathBuf};

#[derive(Debug, Clone)]
#[allow(dead_code)]
struct Record {
    line_num: usize,
    path: String,
    type_field: String,
    size: String,
    descendants: String,
}

/// Parse a CSV line with quoted fields (handles commas inside quotes)
fn parse_csv_line(line: &str) -> Vec<String> {
    let mut fields = Vec::new();
    let mut current = String::new();
    let mut in_quotes = false;
    let mut chars = line.chars().peekable();

    while let Some(c) = chars.next() {
        match c {
            '"' => {
                if in_quotes {
                    // Check for escaped quote ""
                    if chars.peek() == Some(&'"') {
                        current.push('"');
                        chars.next();
                    } else {
                        in_quotes = false;
                    }
                } else {
                    in_quotes = true;
                }
            }
            ',' if !in_quotes => {
                fields.push(current.clone());
                current.clear();
            }
            _ => current.push(c),
        }
    }
    fields.push(current);
    fields
}

fn parse_uffs_output(filepath: &Path) -> Vec<Record> {
    let file = match File::open(filepath) {
        Ok(f) => f,
        Err(e) => {
            eprintln!("Error opening {}: {}", filepath.display(), e);
            return Vec::new();
        }
    };
    let reader = BufReader::new(file);
    let mut records = Vec::new();

    for (line_num, line_result) in reader.lines().enumerate() {
        let line = match line_result {
            Ok(l) => l,
            Err(_) => continue,
        };
        if line.is_empty() {
            continue;
        }

        // Skip header line
        if line.starts_with("\"Path\"") || line.starts_with("Path") {
            continue;
        }

        let parts = parse_csv_line(&line);

        // CSV format: Path, Name, Path Only, Size, Size on Disk, Created, Last Written, Last Accessed, Descendants, ...
        // Index:      0     1     2          3     4             5        6             7              8
        if parts.len() >= 9 {
            records.push(Record {
                line_num: line_num + 1,
                path: parts[0].clone(),
                type_field: String::new(), // Not in this format
                size: parts[3].clone(),
                descendants: parts[8].clone(),
            });
        }
    }
    records
}

fn normalize_path(path: &str) -> String {
    path.to_lowercase().replace('\\', "/").trim_end_matches('/').to_string()
}

fn compare_outputs(cpp_file: &Path, rust_file: &Path) {
    println!("\n{}", "=".repeat(70));
    println!("Comparing:");
    println!("  C++:  {}", cpp_file.display());
    println!("  Rust: {}", rust_file.display());
    println!("{}", "=".repeat(70));

    let cpp_records = parse_uffs_output(cpp_file);
    let rust_records = parse_uffs_output(rust_file);

    println!("\nRecord counts:");
    println!("  C++:  {:>12}", format_num(cpp_records.len()));
    println!("  Rust: {:>12}", format_num(rust_records.len()));
    let diff = rust_records.len() as i64 - cpp_records.len() as i64;
    println!("  Diff: {:>+12}", diff);

    // Build path maps
    let cpp_paths: HashMap<String, &Record> = cpp_records
        .iter()
        .map(|r| (normalize_path(&r.path), r))
        .collect();
    let rust_paths: HashMap<String, &Record> = rust_records
        .iter()
        .map(|r| (normalize_path(&r.path), r))
        .collect();

    let cpp_keys: HashSet<_> = cpp_paths.keys().cloned().collect();
    let rust_keys: HashSet<_> = rust_paths.keys().cloned().collect();

    let common: HashSet<_> = cpp_keys.intersection(&rust_keys).cloned().collect();
    let cpp_only: Vec<_> = cpp_keys.difference(&rust_keys).cloned().collect();
    let rust_only: Vec<_> = rust_keys.difference(&cpp_keys).cloned().collect();

    println!("\nPath comparison:");
    println!("  Common paths:  {:>12}", format_num(common.len()));
    println!("  C++ only:      {:>12}", format_num(cpp_only.len()));
    println!("  Rust only:     {:>12}", format_num(rust_only.len()));

    if !cpp_paths.is_empty() {
        let match_pct = common.len() as f64 / cpp_paths.len() as f64 * 100.0;
        println!("  Match rate:    {:>11.4}%", match_pct);
    }

    // Check attribute differences
    let mut attr_diffs = Vec::new();
    for path in &common {
        let cpp_rec = cpp_paths.get(path).unwrap();
        let rust_rec = rust_paths.get(path).unwrap();
        if cpp_rec.size != rust_rec.size || cpp_rec.descendants != rust_rec.descendants {
            attr_diffs.push((
                path.clone(),
                cpp_rec.size.clone(),
                rust_rec.size.clone(),
                cpp_rec.descendants.clone(),
                rust_rec.descendants.clone(),
            ));
        }
    }

    if !attr_diffs.is_empty() {
        println!("\n⚠️  Attribute differences in common paths: {}", attr_diffs.len());
        println!("\nFirst 10 attribute differences:");
        for (path, cpp_size, rust_size, cpp_desc, rust_desc) in attr_diffs.iter().take(10) {
            println!("  {}", path);
            println!("    Size: C++={} vs Rust={}", cpp_size, rust_size);
            println!("    Desc: C++={} vs Rust={}", cpp_desc, rust_desc);
        }
    }

    // Report missing paths
    if !cpp_only.is_empty() {
        println!("\n❌ Paths in C++ but NOT in Rust (first 20):");
        let mut sorted: Vec<_> = cpp_only.iter().collect();
        sorted.sort();
        for p in sorted.iter().take(20) {
            println!("  {}", p);
        }
        if cpp_only.len() > 20 {
            println!("  ... and {} more", cpp_only.len() - 20);
        }
    }

    if !rust_only.is_empty() {
        println!("\n❌ Paths in Rust but NOT in C++ (first 20):");
        let mut sorted: Vec<_> = rust_only.iter().collect();
        sorted.sort();
        for p in sorted.iter().take(20) {
            println!("  {}", p);
        }
        if rust_only.len() > 20 {
            println!("  ... and {} more", rust_only.len() - 20);
        }
    }

    // Perfect match check
    if cpp_only.is_empty() && rust_only.is_empty() && attr_diffs.is_empty() {
        println!("\n✅ PERFECT MATCH - All {} paths match exactly!", format_num(common.len()));
    }
}

fn format_num(n: usize) -> String {
    let s = n.to_string();
    let mut result = String::new();
    for (i, c) in s.chars().rev().enumerate() {
        if i > 0 && i % 3 == 0 {
            result.push(',');
        }
        result.push(c);
    }
    result.chars().rev().collect()
}

fn find_files(dir: &Path, pattern: &str) -> Vec<PathBuf> {
    let mut results = Vec::new();
    if let Ok(entries) = fs::read_dir(dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_file() {
                if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
                    if name.starts_with(pattern) && name.ends_with(".txt") {
                        results.push(path);
                    }
                }
            }
        }
    }
    results.sort();
    results
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: rust-script {} <trial_dir>", args[0]);
        eprintln!();
        eprintln!("Expected files in trial_dir:");
        eprintln!("  - cpp_d.txt (or cpp_<drive>.txt) - C++ reference output");
        eprintln!("  - rust_new_d.txt (or rust_new_<drive>.txt) - Rust with new tree algo");
        std::process::exit(1);
    }

    let trial_dir = Path::new(&args[1]);
    if !trial_dir.exists() {
        eprintln!("Error: Directory not found: {}", trial_dir.display());
        std::process::exit(1);
    }

    let cpp_files = find_files(trial_dir, "cpp_");
    let rust_new_files = find_files(trial_dir, "rust_new_");

    println!("Trial directory: {}", trial_dir.display());
    println!("Found {} C++ files, {} Rust (new) files", cpp_files.len(), rust_new_files.len());

    for cpp_file in &cpp_files {
        if let Some(stem) = cpp_file.file_stem().and_then(|s| s.to_str()) {
            let drive = stem.strip_prefix("cpp_").unwrap_or("");
            let rust_file = trial_dir.join(format!("rust_new_{}.txt", drive));
            if rust_file.exists() {
                compare_outputs(cpp_file, &rust_file);
            } else {
                println!("\n⚠️  No matching Rust file for {}", cpp_file.display());
            }
        }
    }
}

